### ISSUE TITLE
Storm v3

### ISSUE THREAD

**Issue by asdine** [asdine/storm#211]
I have finally found some time to work actively on the next version of Storm (:tada: :tada: :tada:).
The idea is quite old (see https://github.com/asdine/storm/issues/169) but i never really could have found the time and energy to redesign something that complicated (Storm is basically a database now).
Now that i do, let's boostrap the next version !

## A new direction

Storm v2 works fine but it suffers from design decisions that were made at a time when Storm was a simple wrapper to simplify some redondant tasks done with BoltDB. Since then, a lot of awesome features where added thanks to dozens of issues and contributions and turned Storm into a very cool database that can be used to perform complicated requests, take advantage of indexes and much more. The original design is now reaching its limit and requires too much energy for Storm to evolve properly.

That's why i think it should be **rewritten**.
But the goal must remain the same: something **simple to use** so even beginners can use it, **powerful** with a lot of features available out of the box, **flexible** so anyone can customize the behaviour of the various components.
This might sound complicated, but the current version of Storm is already decent in all of these criterias. Decent, not great though.

## Wanted features

Here is an ambitious, non exhaustive list of the features we could have in the next version:

- Centralized, index aware query system
- Support for dynamic data (i.e Maps)
- Typed indexes
- Custom indexes
- Aggregation
- Expiration
- Geo indexes
- Code generation
- Better high level API
- Low level API
- Better documentation

## Other features

This is a list of features we could have but i'm not sure about:

- Abstracting the low level storage (aka Support for BuntDB, Badger, etc)
- Search (Bleve)

## Design evolution

*I will keep updating this post as the design evolves*

- Brainstorm about the scan system: [gist](https://gist.github.com/asdine/21f258dfbfdf6670e885699ce3e528d2)
- [Draft of the read pipeline](https://paper.dropbox.com/doc/Storm-v3-read-pipeline-design-1aUkubCxAWDF6W0IrzEEZ?_tk=share_copylink)
- [Second iteration](https://paper.dropbox.com/doc/Storm-v3-design-Second-iteration-DpSa2DX9AsbPLSQkkJ2Jb?_tk=share_copylink) (_2018/04/09_)

## Help wanted

Any comment, feature request, remark, contribution is welcome!

**Comment by PaluMacil** [comment]
I love the idea of bleve support.

Does "Aggregation" or "dynamic data" point somewhat to a resultset format that would also perhaps fit "table joins? And even if so, do you think joined tables would be too complex for Storm?

**Comment by bep** [comment]
I heard that Oracle also started out as a simple database wrapper :-) 

I know your brainstorm/gist is a "work in progress" and just started, and I like what I see, but I think you need to sketch some more of the core flows in Storm (not just reading), including the input side (saves). This will tell you how to handle IDs etc.

**Comment by asdine** [comment]
> Does "Aggregation" or "dynamic data" point somewhat to a resultset format that would also perhaps fit "table joins? And even if so, do you think joined tables would be too complex for Storm?

By aggregation i was talking about joins yes.
I'm currently trying to design a read pipeline that would allow to heavily customize the queries. Each stage of the pipeline should be extensible.
Ideally i would like users to be able to do something like this:
- `SELECT MAX(field)...`
- `SELECT .... WHERE .... GROUP BY field`
- `SELECT .... WHERE field = (SELECT ...)`
-  Joins

The goal is not to map every SQL functions but to have a pipeline that ultimately allows the user to be able to:
- apply transformations
- aggregate/join queries
- perform multiple actions with the same result set

If all the stages are customizable (by passing a function or a custom implementation) that would be very powerful.

By `dynamic data` i was thinking about map support. Currently, Storm is very dependant on structs. With map support we could create buckets with true schemaless data and with indexing support.

> I know your brainstorm/gist is a "work in progress" and just started, and I like what I see, but I think you need to sketch some more of the core flows in Storm (not just reading), including the input side (saves). This will tell you how to handle IDs etc.

Yes totally. i didn't start thinking about writes yet, but yeah writes should be discussed heavily too.

**Comment by asdine** [comment]
Here is a draft of the read pipeline: https://paper.dropbox.com/doc/Storm-v3-read-pipeline-design-1aUkubCxAWDF6W0IrzEEZ?_tk=share_copylink

**Comment by asdine** [comment]
Here is a paper of the second iteration of the design: https://paper.dropbox.com/doc/Storm-v3-design-Second-iteration-DpSa2DX9AsbPLSQkkJ2Jb?_tk=share_copylink

**Comment by gedw99** [comment]
here is a graphql db 
https://github.com/vilterp/treesql
the way it keeps the overhead for the developer low is pretty neat i felt.
does live queries and a graphql style / syntax

**Comment by gedw99** [comment]
Alot of what you are talking about fits the architecture of a graphql system.
For example resolvers for joins / Aggregations.
That's why I sent the link to treesql as it's a type of graphql.

But raising the abstraction your less dependent of the fundamentals of boltdb or another db for that matter.

It also make it easier to strap in bleve and other aspects.

A perfect example of this approach is dgraph. 

Also with graphql you can codegen the golang code from the schema.
This removes the current reflection issues.

I really like storm and using it with protobufs. The graoql layer above but integrated would give the things you want perhaps.
There are already some excellent graphql golang libs with codegen now

**Comment by gedw99** [comment]
That's a good idea.
Is it so you can wuery your global distributed store ?

**Comment by asdine** [comment]
I'm currently focusing on abstracting the storage from Storm to allow to easily choose between Bolt, Badger, memory, etc. I'm not sure this will fit well with a remote storage though, as it might potentially require a lot of network round trips for a single query. One thing that is certain is that we will be able to experiment various storage engines

**Comment by gedw99** [comment]
i see your point regarding the issue of requiring many IO calls to etcd.
2 ideas:
1.
If a GRPC frontend is put in with a query language that is passed over the
grpc, then all Io calls to the DB happen on the etcd cluster and return the
final query.

2.
The other option is to define your "stored procedures" at the grpc level.
then you dont need a query language.
you just have to write a handler for each "stored procedure" that then
talks to the Etcd cluster.

**Comment by drauschenbach** [comment]
@gedw99 my motivation for using Storm (with Bolt or etcd) is its secondary index capabilities. The potential is to be able to write a record and its indexes in a single transaction, backed by MVCC in BoltDB's case, and backed by quorum+MVCC in etcd's case. It'd be a very powerful combination.

**Comment by gedw99** [comment]
@drauschenbach 
 thanks for adding meat on the bones... Its elegant.
what do you think about my two approaches ?

**Comment by covrom** [comment]
I can try develop in-memory btree indexes. Store it in memory without usage of boltdb, and rebuild on nessesary.

**Comment by covrom** [comment]
I think about remove or rebuild code with usage of reflect, because it is bad for long-time usage. Use of reflection would be in short-style.

**Comment by gedw99** [comment]
@drauschenbach 

I think you have misunderstood the architectural implications. Really.

On the client and server they are both "embedded". 
All i am suggesting is on the server layering a high level API on top. It does not effect the embedded aspect. 
Write now people do this with storm and boltdb in that they write a layer on top ( be it GRPC, JSONRPC, GraphQL). That layer is a "stored procedure" in the sense that its a not chatty but a single call and response.

**Comment by drauschenbach** [comment]
You're right, you lost me. This is an embedded DB library. There is no server. What people do with this library could be a server, or it could be anything else.

**Comment by gedw99** [comment]
@drauschenbach clearly you can use it as embedded OR network service. The server just needs a GRPC wrapper for example. 

Anyway, there is so much confusion as to what v3 design is, and thats because the key architectural functionality wanted is not known. 

DO you want the queries strongly typed or interpreted is a big one ? For example a graphql interpreter.

DO you want to support subscriptions ?

**Comment by gedw99** [comment]
Its seems that a pipeline is wanted.

At least decide then if its Orchestration or Choreography based ?
Both approaches result in a pipeline.
Choreography is event based. 

lots of info on google ....

**Comment by gedw99** [comment]
https://paper.dropbox.com/doc/Storm-v3-design-Second-iteration-DpSa2DX9AsbPLSQkkJ2Jb
This looks really good.

From the blog post, it seems the pipeline results in in memory Materialised iews. Thats perfect a perfect approach. This effectively means you dont need joins right ?

Its seems your going or Orchestration, since you have this example
"
// pipeline
fn1(t1) -> t2 ; fn2(t2) -> t3 ; fn3(t3) -> t4;
"
If i am correct let me know.

If you took the Choreography approach than you would have triggers on tables, and other real or "materialized Views" list for events on tables and then react to them.
This is another way to do the Materilised Views ( aka Aggregations).
Importantly a Choreography approach also allows you to do live queries / subscriptions !
This is basically what NATS offers - Choreography.

The thing i am really curious about is typing... 
I woudl want both typed and untypd queries.
Untyped is like Graphql and you can use the graphQL scheme to do the resolves at runtime using reflection ( ala dgraph)
Typed is where you code gen off your schema and so you dont have to do any reflection. 
Its the same code just one is generated at runtime and the other at designtime !

**Comment by asdine** [comment]
> The thing i am really curious about is typing...
I woudl want both typed and untypd queries.
Untyped is like Graphql and you can use the graphQL scheme to do the resolves at runtime using reflection ( ala dgraph)
Typed is where you code gen off your schema and so you dont have to do any reflection.
Its the same code just one is generated at runtime and the other at designtime !

The approach i'm taking is to decorelate what's stored from what is used to store.
What is stored: Everything written in the engine (Bolt, Badger, IndexedDB, etc) is strongly typed and a schema is there to describe how data is structured.

How it is stored: It depends on the input, it could be generated, it could use reflection, it could come from a map, a struct, it doesn't matter for the engine, helper functions just need to produce the same interface.

> From the blog post, it seems the pipeline results in in memory Materialised iews. Thats perfect a perfect approach. This effectively means you dont need joins right ?

Regarding queries, i'm using materialised views indeed and designing a way to give enough flexibility to be able to process data the way the user wants.
Basically, anything that is stored can be transformed and filtered and returned in any compatible target.

**Comment by asdine** [comment]
My main interrogation is that this design is very far from what Storm is today and that instead of transforming Storm i should simply create a new project. 
One possibility would be that Storm could use this project internally and keep it's simplistic API, while the other project would give more power to the ones that need a strongly typed approach.
One other possibility would be that Storm would have nothing to do with the other project.
And the final possibility is to accept that Storm will evolve a lot and keep working on a v3

**Comment by gedw99** [comment]
Thanks for the response.
Not sure I helped and I think I get the gist of your design intent.

I realised yesterday there is a similar thing in sqllite to what I was
talking about. It has the ability to fire table change notifications, so
that you can do pub sub above it and achieve similar to your intents.

You did not really touch on whether you pub sub / pipeline is orchestration
based. Am pretty certain it is. I need to try and run some test code to see

Anyway it would be needed that these events can be reacted to in the users
program of course. For example you would want to fire an event up your
stack towards a GUI or a data flow plane to let it know of the new records
that have been created / deleted / updated in the materialised view.

I need to dig into the tests and see how it all works !

**Comment by se3000** [comment]
We were recently building a migration system on top of v2 and wanted to be able to decouple the names of types from the bucket that they write to. I actually think this wouldn't be too difficult to add into the current version, but figured it might be worth bringing up here as a possible future feature.

**Comment by nustiueudinastea** [comment]
@se3000 any chance you can share the migration code?

**Comment by asdine** [comment]
Hey everyone :wave: 

It's been a while since this issue has been created and even longer since I had that in mind, but I have finally made a first version and I'd like to introduce it to you.
I have decided to create a new project instead of rewriting Storm, because what I had in mind was totally different and it wouldn't make sense to just increment a major version.

Please welcome [**Genji**](https://github.com/asdine/genji) :tada: 

It's a database that uses an engine system that deleguates the storage to existing K-V stores (it currently supports BoltDB and an in memory engine)
It's not using reflection at all, and relies on code generation.

It's not perfect, I still need to write a lot of tests and improve documentation, but I think it's good enough to compete with Storm on a few use cases.

The goal is to improve the Query engine to include Joins etc. and even allow to write SQL queries that would be compiled to Go code, but I'm not there yet :laughing: 

Don't hesitate creating issues :muscle: 

Thank you for your help :+1:
